google/cloud/storage/client.py

```diff
@@ -144,19 +151,23 @@ def __init__(
                 api_endpoint = client_options.api_endpoint
                 kw_args["api_endpoint"] = api_endpoint
 
-        # Use anonymous credentials and no project when
-        # STORAGE_EMULATOR_HOST or a non-default api_endpoint is set.
-        if (
-            kw_args["api_endpoint"] is not None
-            and _BASE_STORAGE_URI not in kw_args["api_endpoint"]
-        ):
-            if credentials is None:
-                credentials = AnonymousCredentials()
-            if project is None:
-                project = _get_environ_project()
-            if project is None:
-                no_project = True
-                project = "<none>"
+        # If a custom endpoint is set, the client checks for credentials
+        # or finds the default credentials based on the current environment.
+        # Authentication may be bypassed under certain conditions:
+        # (1) STORAGE_EMULATOR_HOST is set (for backwards compatibility), OR
+        # (2) use_auth_w_custom_endpoint is set to False.
+        if kw_args["api_endpoint"] is not None:
```

Comment:
I would recommend if the `api_endpoint` has a hostname like `.googleapis.com` that auth be used; otherwise it does require always supplying `use_auth_w_custom_endpoint` everytime. IIRC this was done in Node.js Storage by @shaffeeullah

Reply:
I see what you mean. It makes sense that we'd want to enforce auth on `.googleapis.com` endpoints. Is that the same case for TPC, Regional Endpoints and PSC? In other words, is it reasonable to bypass auth under those use cases?

If so, would it make more sense to only bypass auth when `STORAGE_EMULATOR_HOST` is set?

Reply:
IIRC we discussed this in Node: to rely on `.googleapis.com` to tell us to enable auth by default; otherwise it's left to the user and defaults to no auth;

Reply:
Got it, thanks Frank! I've also commented in the internal bug asking for more details on the TPC/Regional Endpoints/PSC use cases. In particular, will the users have an option to bypass auth in those scenarios?

@shaffeeullah @ddelgrosso1 In Node, can a user set `useAuthWithCustomEndpoint=false`  if the api_endpoint has a hostname like `.googleapis.com`

Reply:
Synced with Node and Go folks offline. We're doing something quite similar here. Updated the PR description for clarity
- `use_auth_w_custom_endpoint` is only used when a custom endpoint is set in conjunction (L159 and L141-143)
- auth is enforced if it's a default endpoint "https://storage.googleapis.com"
- no extra validations on the custom endpoints

Special note
- `use_auth_w_custom_endpoint` defaults to True in Python
- for backwards compatibility, however, auth is automatically bypassed if `STORAGE_EMULATOR_HOST` is set

Reply:
Thanks Cathy, it's a bit tricky; Thanks for updating documentation.